File: base/ai.cc
#include <iostream>
#include <fstream>
#include <cassert>
#include "base/base.h"
#include "base/ai.h"
#include "base/geometry.h"
#include "absl/strings/str_split.h"
#include <glog/logging.h>

Position dir2vec(const Direction &dir) {
  static const int dx[] = {0, 1, 0, -1};
  static const int dy[] = {1, 0, -1, 0};
  int idx = static_cast<int>(dir);

  return Position(dx[idx], dy[idx]);
}

char direction_to_char(Direction d) {
  switch (d) {
  case Direction::Right:
    return 'D';
  case Direction::Left:
    return 'A';
  case Direction::Up:
    return 'W';
  case Direction::Down:
    return 'S';
  }

  LOG(FATAL) << "Invalid direction " << static_cast<int>(d);
  return '!';
}

// relative p when Right => ??? when dir
Position rotate(Position p, Direction d) {
  switch (d) {
  case Direction::Right:
    return { p.first, p.second };
  case Direction::Up:
    return { -p.second, p.first };
  case Direction::Left:
    return { -p.first, -p.second };
  case Direction::Down:
    return { p.second, -p.first };
  }
  LOG(FATAL) << "Invalid direction " << static_cast<int>(d);
  return Position(-1000000, -1000000);
}

// relative p when dir => ??? when Right
Position rotate_reverse(Position p, Direction dir) {
  Position q = rotate(p, dir);
  q = rotate(q, dir);
  q = rotate(q, dir);
  return q;
}

absl::optional<Direction> char_to_direction(char c) {
  switch(c) {
  case 'D':
    return Direction::Right;
  case 'A':
    return Direction::Left;
  case 'W':
    return Direction::Up;
  case 'S':
    return Direction::Down;
  };

  return absl::nullopt;
}

Worker::Worker(Position pos) {
  current_pos = pos;
}

bool AI::valid_pos(Position target) const {
  int x = target.first;
  int y = target.second;
  return 0 <= x && x < height && 0 <= y && y < width;
}

bool AI::reachable(Position target, const int id) const {
  int x = target.first;
  int y = target.second;
  if (!valid_pos(target))
    return false;

  if (board[x][y] == '#')
    return false;

  auto worker_pos = get_pos(id);
  int wx = worker_pos.first;
  int wy = worker_pos.second;
  for (int cx = std::min(x, wx); cx <= std::max(x, wx); cx++) {
    for (int cy = std::min(y, wy); cy <= std::max(y, wy); cy++) {
      if (board[cx][cy] != '#')
        continue;

      Line l({wx + 0.5, wy + 0.5}, {x + 0.5, y + 0.5});

      Point cell(cx+0.5, cy+0.5);
      if (intersectSP(l, cell))
        return false;

      for(int i = 0; i < 4; i++) {
        const int dx[] = {cx, cx+1, cx+1, cx};
        const int dy[] = {cy, cy, cy+1, cy+1};
        Line edge({(double)dx[i], (double)dy[i]}, {(double)dx[(i+1)%4], (double)dy[(i+1)%4]});

        if (intersectSS(l, edge))
          return false;
      }
    }
  }

  return true;
}

void AI::initialize() {
  std::cin >> height >> width;
  std::string buf;
  while (std::cin >> buf) {
    board.push_back(buf);
  }

  Position worker_pos = std::make_pair(0, 0);
  for (int i = 0; i < height; ++i) {
    for (int j = 0; j < width; ++j) {
      if (board[i][j] == 'W') {
        worker_pos = std::make_pair(i, j);
      } else if (board[i][j] == '#') {
        block_count++;
      } else if (board[i][j] == 'X') {
	spawn_points.emplace_back(i, j);
      } else if (board[i][j] == 'C') {
	cloning_points.emplace_back(i, j);
      }
    }
  }
  filled.resize(height);
  for (int i = 0; i < height; ++i) filled[i].resize(width);

  workers.clear();
  workers.push_back(Worker(worker_pos));

  executed_cmds.resize(1);

  for (auto &p : workers[0].manipulator_range) {
    auto x = workers[0].current_pos.first + p.first;
    auto y = workers[0].current_pos.second + p.second;
    fill_cell(std::make_pair(x, y), 0);
  }

  pickup_booster(0);
}

bool AI::init_turn(const int id) {
  LOG_IF(FATAL, expect_worker_id != id)
    << "command for the worker " << expect_worker_id
    << "must be called but the ID was" << id;

  // Pick up a booster.
  // See Appendix in clone document
  pickup_booster(id);

  return !is_finished();
}

AI::AI() {
  initialize();
}

AI::AI(const std::string filename) {
  std::ifstream in(filename.c_str());
  LOG_IF(FATAL, !in) <<
    "failed to open " << filename;

  std::cin.rdbuf(in.rdbuf());

  initialize();
}

void AI::pickup_booster(const int id) {
  auto p = get_pos(id);


  switch(board[p.first][p.second]) {
  case 'B':
    count_extension += 1;
    break;
  case 'F':
    count_fast += 1;
    break;
  case 'L':
    count_drill += 1;
    break;
  case 'C':
    count_clone += 1;
    break;
  case 'R':
    count_teleport += 1;
    break;
  default:
    break;
  }
  board[p.first][p.second] = '.';

}

bool AI::fill_cell(Position pos, const int id) {
  int x = pos.first;
  int y = pos.second;
  if (!reachable(pos, id)) {
    return false;
  }

  if (!filled[x][y]) {
    filled_count++;
  }
  filled[x][y] = true;

  return true;
}

int AI::get_height() { return height; }
int AI::get_width() { return width; }

int AI::get_block_count() const { return block_count; }

int AI::get_time() { return current_time; }

Position AI::get_pos(const int id) const { return workers[id].current_pos; }

Direction AI::get_dir(const int id) { return workers[id].current_dir; }

int AI::get_filled_count() { return filled_count; }

int AI::get_count_fast() { return count_fast; }
int AI::get_count_drill() { return count_drill; }
int AI::get_count_extension() { return count_extension; }
int AI::get_count_clone() { return count_clone; }
int AI::get_count_teleport() { return count_teleport; }

int AI::get_count_workers() { return workers.size(); }

int AI::get_duration_fast(const int id) { return workers[id].duration_fast; }
int AI::get_duration_drill(const int id) { return workers[id].duration_drill; }

Position AI::get_neighbor(const Direction &dir, const int id) {
  const int dx[] = {0,1, 0,-1};
  const int dy[] = {1,0,-1, 0};
  int idx = static_cast<int>(dir);

  return Position(workers[id].current_pos.first + dx[idx],
                  workers[id].current_pos.second + dy[idx]);
}

std::vector<Position> AI::get_absolute_manipulator_positions(const int id) {

  Position self = get_pos(id);
  std::vector<Position> ret;

  for (Position&p : workers[id].manipulator_range) {
    Position mani = rotate(p, get_dir(id));
    mani.first += self.first;
    mani.second += self.second;
    ret.push_back(mani);
  }
  return ret;
}

std::vector<Position> AI::rotated_manipulator_positions(bool is_cw, const int id) {
  Position self = get_pos(id);
  std::vector<Position> ret;

  for (Position&p : workers[id].manipulator_range) {
    Direction cur_dir = get_dir(id);
    int ndir = (static_cast<int>(cur_dir) + (is_cw ? 1 : 3)) % 4;
    Direction next_dir = static_cast<Direction>(ndir);

    Position mani = rotate(p, next_dir);
    mani.first += self.first;
    mani.second += self.second;

    if(reachable(mani, id))
      ret.push_back(mani);
  }
  return ret;
}

std::vector<Position> AI::moved_manipulator_positions(const Direction& d, const int id) const {
  Position self = get_pos(id);
  std::vector<Position> ret;

  const auto v = dir2vec(d);

  for (const Position &p : workers[id].manipulator_range) {
    const Position mani = {self.first + p.first + v.first,
			   self.second + p.second + v.second};
    
    if(reachable(mani, id)) {
      ret.push_back(mani);
    }
  }

  return ret;
}

void AI::next_turn(const int id) {
  if (id == (int)workers.size()-1)
    current_time++;

  expect_worker_id += 1;
  expect_worker_id %= workers.size();

  if (workers[id].duration_drill > 0)
    workers[id].duration_drill--;
  if (workers[id].duration_fast > 0)
    workers[id].duration_fast--;
}

void AI::push_command(Command cmd, const int id) {
  LOG_IF(FATAL, id >= (int)executed_cmds.size()) << id << " is out of range " << workers.size();

  executed_cmds[id].push_back(cmd);
}

bool AI::try_move(const Direction &dir, const int id) {
  Position next_pos = get_neighbor(dir, id);

  // Check validity
  if (!valid_pos(next_pos))
    return false;

  if (board[next_pos.first][next_pos.second] == '#' && workers[id].duration_drill == 0)
    return false;

  return true;
}

bool AI::move_body(const Direction &dir, const int id) {
  // Checks if this is a valid move
  if(!try_move(dir, id))
    return false;

  auto pos = get_neighbor(dir);
  // Move
  workers[id].current_pos = pos;

  // Check drill
  if (board[pos.first][pos.second] == '#' &&
      0 < workers[id].duration_drill) {
    board[pos.first][pos.second] = '.';
    block_count--;
  }

  // Pick up items
  for(auto p: get_absolute_manipulator_positions(id)) {
    fill_cell(p, id);
  }

  return true;
}

bool AI::move(const Direction &dir, const int id) {
  if(!init_turn(id))
    return false;

  if (!move_body(dir, id))
    return false;

  // When FAST booster is enabled, move twice
  if (workers[id].duration_fast > 0) {
    move_body(dir, id);
  }

  Command cmd = {CmdType::Move, dir};
  push_command(cmd, id);

  // Update time
  next_turn(id);

  return true;
}

bool AI::turn_CW(const int id) {
  if(!init_turn(id))
    return false;

  workers[id].current_dir =
    static_cast<Direction>( ( static_cast<int>(workers[id].current_dir) + 1 ) % 4 );

  for(auto p: get_absolute_manipulator_positions()) {
    fill_cell(p);
  }

  Command cmd = {CmdType::TurnCW};
  push_command(cmd, id);

  next_turn(id);

  return true;
}

bool AI::turn_CCW(const int id) {
  if(!init_turn(id))
    return false;

  workers[id].current_dir =
    static_cast<Direction>( ( static_cast<int>(workers[id].current_dir) + 3 ) % 4 );

  for(auto p: get_absolute_manipulator_positions()) {
    fill_cell(p);
  }

  Command cmd = {CmdType::TurnCCW};
  push_command(cmd, id);

  next_turn(id);

  return true;
}

bool AI::use_fast_wheel(const int id) {
  if(!init_turn(id))
    return false;

  if (count_fast == 0)
    return false;

  workers[id].duration_fast = Worker::DURATION_FAST_MAX;
  count_fast--;

  Command cmd = {CmdType::UseFastWheel};
  push_command(cmd, id);

  next_turn(id);
  return true;
}

bool AI::use_drill(const int id) {
  if(!init_turn(id))
    return false;
  if (count_drill == 0)
    return false;

  workers[id].duration_drill = Worker::DURATION_DRILL_MAX;
  count_drill--;

  Command cmd = {CmdType::UseDrill};
  push_command(cmd, id);

  next_turn(id);

  return true;
}

bool AI::use_extension(const int dx, const int dy, const int id) {
  if(!init_turn(id))
    return false;

  if (count_extension == 0)
    return false;

  for (auto m: workers[id].manipulator_range) {
    if (Position(dx, dy) == m) {
      return false;
    }
  }

  bool can_use = false;

  for(auto m: workers[id].manipulator_range) {
    if (std::abs(m.first - dx) + std::abs(m.second - dy) == 1) {
      can_use = true;
    }
  }

  if(!can_use)
    return false;

  workers[id].manipulator_range.push_back( rotate_reverse({dx, dy}, get_dir(id)) );
  count_extension--;

  // Fill a cell visited by a new manipulator.
  // Update all manipulator's cell by get_absolute_manipulator_positions() because I'm lazy.
  for(auto p: get_absolute_manipulator_positions(id)) {
    fill_cell(p, id);
  }

  Command cmd = {CmdType::UseExtension, Direction::Right /* dymmy */, dx, dy};
  push_command(cmd, id);

  next_turn(id);

  return true;
}

bool AI::use_clone(const int id) {
  if(!init_turn(id))
    return false;
  if (count_clone == 0)
    return false;

  auto p = get_pos(id);
  if (board[p.first][p.second] != 'X')
    return false;

  count_clone--;

  // Create a child worker
  Worker child(p);
  workers.push_back(child);

  executed_cmds.push_back(std::vector<Command>());

  Command cmd = {CmdType::UseClone};
  push_command(cmd, id);

  next_turn(id);
  return true;
}

bool AI::install_beacon(const int id) {
  if(!init_turn(id))
    return false;

  auto p = get_pos(id);
  if (board[p.first][p.second] == 'X' ||
      board[p.first][p.second] == 'b') {
    return false;
  }

  if (count_teleport == 0)
    return false;

  count_teleport--;
  board[p.first][p.second] = 'b';
  beacon_pos.insert(p);

  push_command({CmdType::InstallBeacon}, id);

  next_turn(id);
  return true;
}

bool AI::jump_to_beacon(Position dst, const int id) {
  if(!init_turn(id))
    return false;

  if (beacon_pos.find(dst) == beacon_pos.end()) {
    return false;
  }

  workers[id].current_pos = dst;

  Command com = {CmdType::JumpToBeacon, Direction::Right /* dummy*/, dst.first, dst.second};
  push_command(com, id);

  next_turn(id);
  return true;
}

bool AI::do_command(Command cmd, const int id) {
  switch (cmd.type) {
  case CmdType::Move:
    return move(cmd.dir, id);
  case CmdType::TurnCW:
    return turn_CW(id);
  case CmdType::TurnCCW:
    return turn_CCW(id);
  case CmdType::UseExtension:
    return use_extension(cmd.x, cmd.y, id);
  case CmdType::UseFastWheel:
    return use_fast_wheel(id);
  case CmdType::UseDrill:
    return use_drill(id);
  case CmdType::UseClone:
    return use_clone(id);
  case CmdType::InstallBeacon:
    return install_beacon(id);
  case CmdType::JumpToBeacon:
    return jump_to_beacon(Position(cmd.x, cmd.y), id);
  }

  LOG(FATAL) << "BUG?: unknown command name: " << static_cast<int>(cmd.type);
}

bool AI::is_finished() {
  return get_filled_count() + block_count == height * width;
}

void print_command(struct Command cmd, int height) {
  std::string s = "";

  switch (cmd.type) {
  case CmdType::Move:
    switch (cmd.dir) {
    case Direction::Right:
      s = "D";
      break;
    case Direction::Down:
      s = "S";
      break;
    case Direction::Left:
      s = "A";
      break;
    case Direction::Up:
      s = "W";
      break;
    }
    break;
  case CmdType::TurnCW:
    s = "E";
    break;
  case CmdType::TurnCCW:
    s = "Q";
    break;
  case CmdType::UseExtension:
    // Convert coordinate: (dx, dy) -> (dy, -dx)
    s = "B(" + std::to_string(cmd.y) + "," + std::to_string(-cmd.x) + ")";
    break;
  case CmdType::UseFastWheel:
    s = "F";
    break;
  case CmdType::UseDrill:
    s = "L";
    break;
  case CmdType::UseClone:
    s = "C";
    break;
  case CmdType::InstallBeacon:
    // Convert coordinate: (dx, dy) -> (dy, -dx)
    s = "R";
    break;
  case CmdType::JumpToBeacon:
    s = "T(" + std::to_string(cmd.y) + "," + std::to_string(height - cmd.x) + ")";
    break;
  default:
    LOG(FATAL) << "BUG?: unknown command name: " << static_cast<int>(cmd.type);
  }

  LOG_IF(FATAL, s.size() == 0) << "BUG: empty command: " << static_cast<int>(cmd.type);

  std::cout << s;
}

void AI::print_commands() {
  for(int i = 0; i < (int) executed_cmds.size(); ++i) {
    if (i != 0)
      std::cout << "#";

    for (auto cmd: executed_cmds[i]) {
      print_command(cmd, height);
    }
  }
  std::cout << std::endl;
}

void AI::dump_state() {
  std::cerr << "time: " << current_time << ", filled_count: " << filled_count << std::endl;

  std::set<Position> ws;

  for (uint32_t i = 0; i < workers.size(); i++) {
    auto p = get_pos(i);
    ws.insert(p);

    std::cerr << "pos[" << i << "]: ("
              << p.first << ", " << p.second << ")"
              << std::endl;
    std::cerr << "duration: drill=" << workers[i].duration_drill
              << ", fast: " << workers[i].duration_fast << std::endl;
  }

  for(int i = 0; i < height; ++i) {
    for(int j = 0; j < width; ++j) {
      if (ws.find(std::make_pair(i, j)) != ws.end()) {
        std::cerr << "W";
      } else if(filled[i][j]) {
        std::cerr << "V";
      } else {
        std::cerr << board[i][j];
      }
    }
    std::cerr << std::endl;;
  }
}

File: cympfh/src/ai.rs
#![allow(dead_code)]

use std::cmp::{min, max};
use std::collections::HashSet;

use crate::geo::{Line, Point, intersect_ss, intersect_sp};

#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
pub enum Direction {
  Right,
  Down,
  Left,
  Up,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Position(pub isize, pub isize);

#[derive(Debug, Clone, Copy)]
pub enum Command {
  Move(Direction),
  Rotate(bool),
  Boost
}

#[derive(Debug, Clone, Hash, PartialEq, Eq)]
pub struct Worker {
    pub current_dir: Direction,
    pub current_pos: Position,
    pub manipulator_range: Vec<Position>,
    pub duration_drill: usize,
    pub duration_fast: usize,
}

impl Position {
    fn rotate(&self, dir: Direction) -> Position {
        match dir {
            Direction::Right => self.clone(),
            Direction::Up => Position(-self.1, self.0),
            Direction::Left => Position(-self.0, -self.1),
            Direction::Down => Position(self.1, -self.0),
        }
    }
    fn rotate_reverse(&self, dir: Direction) -> Position {
        self.rotate(dir).rotate(dir).rotate(dir)
    }
}

impl Direction {
    pub fn to_pos(&self) -> Position {
        match self {
            Direction::Right => Position(0, 1),
            Direction::Down => Position(1, 0),
            Direction::Left => Position(0, -1),
            Direction::Up => Position(-1, 0),
        }
    }
    pub fn cw(&self) -> Direction {
        match self {
            Direction::Right => Direction::Down,
            Direction::Down => Direction::Left,
            Direction::Left => Direction::Up,
            Direction::Up => Direction::Right,
        }
    }
    pub fn ccw(&self) -> Direction {
        match self {
            Direction::Right => Direction::Up,
            Direction::Down => Direction::Right,
            Direction::Left => Direction::Down,
            Direction::Up => Direction::Left,
        }
    }
}

impl Worker {
    fn new(p: Position) -> Self {
        Worker {
            current_dir: Direction::Right,
            current_pos: p,
            manipulator_range: vec![
                Position(0, 0),
                Position(1, 1),
                Position(0, 1),
                Position(-1, 1)
            ],
            duration_drill: 0,
            duration_fast: 0,
        }
    }
}

#[derive(Debug, Clone)]
pub struct AI {
    pub current_time: usize,
    pub height: usize,
    pub width: usize,
    pub count_fast: usize,
    pub count_drill: usize,
    pub count_extension: usize,
    pub count_clone: usize,
    pub count_teleport: usize,
    pub filled_count: usize,
    pub block_count: usize,
    pub workers: Vec<Worker>,
    pub beacons: HashSet<Position>,
    pub executed_cmds: Vec<Vec<String>>,
    pub board: Vec<Vec<char>>,
    pub filled: Vec<Vec<bool>>,
}

impl AI {

    pub fn new(h: usize, w: usize, board: Vec<Vec<char>>) -> Self {
        let w_pos = {
            let mut pos = Position(0, 0);
            for i in 0..h {
                for j in 0..w {
                    if board[i][j] == 'W' {
                        pos = Position(i as isize, j as isize);
                    }
                }
            }
            pos
        };
        let block_count = board.iter().map(|row|
                  row.iter().filter(|&c| *c == '#').count()).sum();
        AI {
            current_time: 0,
            height: h,
            width: w,
            count_fast: 0,
            count_drill: 0,
            count_extension: 0,
            count_clone: 0,
            count_teleport: 0,
            filled_count: 0,
            block_count: block_count,
            workers: vec![Worker::new(w_pos)],
            executed_cmds: vec![vec![]],
            beacons: HashSet::new(),
            board: board,
            filled: vec![vec![false; w]; h],
        }
    }

    fn valid_pos(&self, p: &Position) -> bool {
        let Position(x, y) = *p;
        0 <= x && x < self.height as isize && 0 <= y && y < self.width as isize
    }

    pub fn reachable(&self, idx: usize, p: &Position) -> bool {
        if !self.valid_pos(p) { return false; }
        let Position(x, y) = *p;
        if self.board[x as usize][y as usize] == '#' { return false; }
        let Position(wx, wy) = self.workers[idx].current_pos;
        for cx in min(x, wx)..=max(x, wx) {
          for cy in min(y, wy)..=max(y, wy) {
            if self.board[cx as usize][cy as usize] != '#' { continue }
            let l = Line(Point(wx as f64 +0.5, wy as f64 + 0.5),
                         Point(x as f64 + 0.5, y as f64 + 0.5));
            let cell = Point(cx as f64 + 0.5, cy as f64 + 0.5);
            if intersect_sp(&l, &cell) { return false; }

            let corners = vec![
              Point(cx as f64, cy as f64),
              Point(cx as f64 + 1.0, cy as f64),
              Point(cx as f64 + 1.0, cy as f64 + 1.0),
              Point(cx as f64, cy as f64 + 1.0),
            ];
            for i in 0..4 {
              let edge = Line(corners[i].clone(), corners[(i + 1) % 4].clone());
              if intersect_ss(&l, &edge) { return false; }
            }
          }
        }
        true
    }

    pub fn fill_cell(&mut self, idx: usize, p: &Position) -> bool {
        let Position(x, y) = *p;
        if !self.valid_pos(p) { return false; }
        if !self.reachable(idx, p) { return false; }

        // pick up when item is on body,
        if self.workers[idx].current_pos == *p {  // when on body
          match self.board[x as usize][y as usize] {
              'B' => { self.count_extension += 1; },
              'F' => { self.count_fast += 1; },
              'L' => { self.count_drill += 1; },
              'C' => { self.count_clone += 1; },
              'R' => { self.count_teleport += 1; },
              _ => {}
          }
          self.board[x as usize][y as usize] = '.';
        }

        if self.filled[x as usize][y as usize] { return false; }
        self.filled[x as usize][y as usize] = true;
        self.filled_count += 1;
        true
    }

    pub fn get_neighbor(&self, idx: usize, d: Direction) -> Position {
        let Position(x, y) = self.workers[idx].current_pos;
        let Position(dx, dy) = d.to_pos();
        Position(x + dx, y + dy)
    }

    pub fn turn_cw(&mut self, idx: usize) -> bool {
        self.workers[idx].current_dir = match self.workers[idx].current_dir {
            Direction::Right => Direction::Down,
            Direction::Down => Direction::Left,
            Direction::Left => Direction::Up,
            Direction::Up => Direction::Right,
        };
        self.executed_cmds[idx].push(String::from("E"));
        for &p in self.get_absolute_manipulator_positions(idx).iter() {
            self.fill_cell(idx, &p);
        }
        self.next_turn();
        true
    }

    pub fn turn_ccw(&mut self, idx: usize) -> bool {
        self.workers[idx].current_dir = match self.workers[idx].current_dir {
            Direction::Right => Direction::Up,
            Direction::Up => Direction::Left,
            Direction::Left => Direction::Down,
            Direction::Down => Direction::Right,
        };
        self.executed_cmds[idx].push(String::from("Q"));
        for &p in self.get_absolute_manipulator_positions(idx).iter() {
            self.fill_cell(idx, &p);
        }
        self.next_turn();
        true
    }

    pub fn get_absolute_manipulator_positions(&self, idx: usize) -> Vec<Position> {
        let Position(x, y) = self.workers[idx].current_pos;
        let mut ret = vec![];
        for &p in self.workers[idx].manipulator_range.iter() {
            let q = p.rotate(self.workers[idx].current_dir);
            ret.push(Position(x + q.0, y + q.1));
        }
        ret
    }

    pub fn rotated_manipulator_positions(&self, idx: usize, cw: bool) -> Vec<Position> {
        let mut ret = vec![];
        let Position(wx, wy) = self.workers[idx].current_pos;
        let ndir = if cw {
          self.workers[idx].current_dir.cw()
        } else {
          self.workers[idx].current_dir.ccw()
        };
        for &p in self.workers[idx].manipulator_range.iter() {
            let mani = p.rotate(ndir);
            let q = Position(mani.0 + wx, mani.1 + wy);
            if self.reachable(idx, &q) {
                ret.push(q);
            }
        }
        ret
    }

    pub fn moved_manipulator_positions(&self, idx: usize, d: Direction) -> Vec<Position> {
        let mut ret = vec![];
        if !self.try_move(idx, d) { return ret }
        let Position(wx, wy) = self.workers[idx].current_pos;
        let v = d.to_pos();
        for &p in self.workers[idx].manipulator_range.iter() {
            let q = Position(wx + p.0 + v.0, wy + p.1 + v.1);
            if self.reachable(idx, &q) {
                ret.push(q);
            }
        }
        ret
    }

    fn next_turn(&mut self) {
        self.current_time += 1;
        for i in 0..self.workers.len() {
            if self.workers[i].duration_drill > 0 {
                self.workers[i].duration_drill -= 1;
            }
            if self.workers[i].duration_fast > 0 {
                self.workers[i].duration_fast -= 1;
            }
        }
    }

    fn try_move(&self, idx: usize, dir: Direction) -> bool {
        let next_pos = self.get_neighbor(idx, dir);
        if !self.valid_pos(&next_pos) {
            return false;
        }
        let Position(x, y) = next_pos;
        if self.board[x as usize][y as usize] == '#' && self.workers[idx].duration_drill == 0 {
            return false;
        }
        true
    }

    fn move_body(&mut self, idx: usize, dir: Direction) -> bool {
        // validation
        if !self.try_move(idx, dir) { return false; }
        // move
        self.workers[idx].current_pos = self.get_neighbor(idx, dir);

        let Position(wx, wy) = self.workers[idx].current_pos;
        if self.board[wx as usize][wy as usize] == '#' {
          self.board[wx as usize][wy as usize] = '.';
          self.block_count -= 1;
        }

        // fill && pick up
        for &p in self.get_absolute_manipulator_positions(idx).iter() {
            self.fill_cell(idx, &p);
        }
        true
    }

    pub fn mv(&mut self, idx: usize, dir: Direction) -> bool {
        if !self.move_body(idx, dir) { return false; }
        // when FAST
        if self.workers[idx].duration_fast > 0 {
            self.move_body(idx, dir);
        }
        // push command
        self.executed_cmds[idx].push(match dir {
            Direction::Up => String::from("W"),
            Direction::Down => String::from("S"),
            Direction::Left => String::from("A"),
            Direction::Right => String::from("D"),
        });
        self.next_turn();
        true
    }

    pub fn is_finished(&self) -> bool {
        self.filled_count + self.block_count == self.height * self.width
    }

    pub fn print_commands(&self) -> String {
        self.executed_cmds.iter().map(|cmds| cmds.join("")).collect::<Vec<String>>().join("#")
    }

    pub fn use_fast_wheel(&mut self, idx: usize) -> bool {
        if self.count_fast == 0 {
            return false;
        }
        self.workers[idx].duration_fast = 50;
        self.count_fast -= 1;
        self.executed_cmds[idx].push(String::from("F"));
        true
    }

    pub fn use_drill(&mut self, idx: usize) -> bool {
        if self.count_drill == 0 {
            return false;
        }
        self.workers[idx].duration_drill = 30;
        self.count_drill -= 1;
        self.executed_cmds[idx].push(String::from("L"));
        true
    }

    pub fn use_clone(&mut self, idx: usize) -> bool {
        if self.count_clone == 0 {
            return false;
        }

        let Position(wx,wy) = self.workers[idx].current_pos;
        if self.board[wx as usize][wy as usize] != 'X' {
            return false;
        }

        self.count_clone -= 1;

        self.workers.push(Worker::new(Position(wx, wy)));

        self.executed_cmds.push(vec![]);
        self.executed_cmds[idx].push(String::from("C"));
        true
    }

    pub fn install_beacon(&mut self, idx: usize) -> bool {
        if self.count_teleport == 0 {
            return false;
        }
        let Position(wx, wy) = self.workers[idx].current_pos;
        if self.board[wx as usize][wy as usize] == 'X' || self.board[wx as usize][wy as usize] == 'b' {
            return false;
        }

        self.count_teleport -= 1;
        self.board[wx as usize][wy as usize] = 'b';

        self.beacons.insert(Position(wx,wy));

        self.executed_cmds[idx].push(String::from("R"));

        true
    }

    pub fn jump_to_beacon(&mut self, idx: usize, dst: &Position) -> bool {
        if !self.beacons.contains(dst) {
            return false;
        }

        self.workers[idx].current_pos.0 = dst.0;
        self.workers[idx].current_pos.1 = dst.1;

        self.executed_cmds[idx].push(format!("T({},{})",dst.0, dst.1));
        true
    }

    pub fn nop(&mut self, idx: usize) -> bool {
        self.executed_cmds[idx].push(String::from("Z"));
        true
    }

    // return relative positions
    pub fn extension_positions(&self, idx: usize) -> Vec<Position> {
        let Position(wx, wy) = self.workers[idx].current_pos;
        let mybody = self.get_absolute_manipulator_positions(0);
        let is_my_body = |p: &Position| { mybody.iter().any(|&q| *p == q) };
        let mut ret = vec![];
        for &Position(x, y) in mybody.iter() {
            for p in vec![Position(x + 1, y), Position(x - 1, y), Position(x, y + 1), Position(x, y - 1)] {
                if !is_my_body(&p) {
                    ret.push(Position(p.0 - wx, p.1 - wy));
                }
            }
        }
        ret
    }

    // p is relative
    pub fn use_extension(&mut self, idx: usize, p: Position) -> bool {
        if self.count_extension == 0 {
            return false;
        }
        if !self.extension_positions(idx).iter().any(|&q| q == p) {
          return false;
        }
        self.count_extension -= 1;
        self.executed_cmds[idx].push(format!("B({},{})", p.1, -p.0));
        {
          let q = p.rotate_reverse(self.workers[idx].current_dir);
          self.workers[idx].manipulator_range.push(q);
        }
        for &p in self.get_absolute_manipulator_positions(idx).iter() {
            self.fill_cell(idx, &p);
        }
        true
    }
}
